packages/runtime/container-runtime/src/summary/runningSummarizer.ts

```diff
@@ -664,6 +661,139 @@ export class RunningSummarizer implements IDisposable {
 		});
 	}
 
+	/**
+	 * Tries to summarize with dynamic retries, i.e., retry is based on the failure params and error received.
+	 * For instance, if summarize failed due to a server error that has `retryAfterSeconds` set, wait for that duration
+	 * and retry summarize. Similarly, if summary generation failed but the error is retriable, retry before bailing out.
+	 */
+	private async trySummarizeWithDynamicRetries(
```

Comment:
Note this change seems rather complex. Dynamic retries is something new, I wonder what sort of impact it will have on the whole system.

Reply:
How will the summarizer perform in the worst-case scenario? Is that something that will cause us grief down the road? It still feels like we need to think about state across multiple refreshes. I'm not sure if this puts us back to square one. I'm rather concerned that we don't always want to retry on the same client. This should be offered as a solution to only certain problems. I'm concerned that if state is bad on a summarizer client, that we will summarize and create even worse state for the next summarization. These are very hard things to test IMO.

Reply:
I am not sure I understand all your concerns.

> This should be offered as a solution to only certain problems.
I am not claiming this solves all problems. This only solves a certain class of issues where the call site thinks retries can help. It's optimitic and summarization can still fail. That will result in closing summarizer and doing summarization from a fresh state.

> How will the summarizer perform in the worst-case scenario? Is that something that will cause us grief down the road?
This change does not introduce any new issues. In fact, it tries to address issues that may be introduced by the recent changes where we only do one attempt. So, it should only make the system better. I can't predict whether this will cause us "grief down the down". It's behind a feature flag and as I mentioned in the description, I will add testing for it. It will also be tested in stress test. I am not sure what else we can do.

> I'm concerned that if state is bad on a summarizer client, that we will summarize and create even worse state for the next summarization.
I do not understand this concern. How does this change result in summarizing in bad state?

> It still feels like we need to think about state across multiple refreshes.
Sure, that can still be added incrementally if needed. I don't think it is needed so I am not adding it here. I would rather look at the data after this and then make a call. If you believe that is needed now, feel free to experiment with it. It can be done indepent of this.